query: introduce space optimized BranchesRepos

tsenart opened the pull request:
This PR introduces a space optimized alternative to `query.RepoBranches`. It's name is `query.BranchesRepos` and it is a slice of `(Branch, Repo IDs bitmap)`. This is part of the effort to reduce memory spikes and OOMs in the sourcegraph frontend.

Commit:
all: space optimized repo sets (WIP)

tsenart commented:
FYI @keegancsmith, working on this

tsenart commented:
Marshalling benchmarks looking good:

```
name                        time/op
RepoBranches_Encode/Set-12   655µs ±43%
RepoBranches_Encode/IDs-12  2.02µs ± 2%
RepoBranches_Decode/Set-12   917µs ± 1%
RepoBranches_Decode/IDs-12  15.4µs ± 1%

name                        bytes
RepoBranches_Encode/Set-12   344kB ± 0%
RepoBranches_Encode/IDs-12  1.75kB ± 0%
RepoBranches_Decode/Set-12
RepoBranches_Decode/IDs-12

name                        alloc/op
RepoBranches_Encode/Set-12   344kB ± 0%
RepoBranches_Encode/IDs-12  2.78kB ± 0%
RepoBranches_Decode/Set-12  1.44MB ± 0%
RepoBranches_Decode/IDs-12  11.1kB ± 0%

name                        allocs/op
RepoBranches_Encode/Set-12    2.00 ± 0%
RepoBranches_Encode/IDs-12    6.00 ± 0%
RepoBranches_Decode/Set-12     380 ± 0%
RepoBranches_Decode/IDs-12     194 ± 0%
```

keegancsmith committed:
only override Repository.ID if RawConfig.repoid is set

keegancsmith committed:
shards: handle IDs in selectRepoSet

This code is untested and is not matching the style of the rest of the
branch. This is purely keegan hacking to understand.

Commit:
test handling of RepoBranches.IDs in indexData.simplify

Commit:
test single repo indexData.simplify of RepoBranches

Commit:
support RepoBranches.IDs in newMatchTree

Commit:
test RepoBranches.IDs in TestFilteringShardsByRepoSet

Commit:
add nil checks in selectRepoSet

Commit:
TestSimplifyRepoSet_IDs

Commit:
TestRepoSet IDs

Commit:
estimate setSize for RepoSet with IDs

tsenart commented on eval_test.go:
```diff
@@ -189,37 +214,124 @@ func TestSimplifyRepo(t *testing.T) {
-	some := &query.RepoBranches{Set: map[string][]string{"bar": {"branch1"}}}
-	none := &query.Repo{"banana"}
+	none := &query.Repo{Pattern: "banana"}
+	for _, tc := range []struct {
```
I could re-work tests in this PR to add an independent test rather than refactoring them into table driven tests which makes the diff harder to read.

Commit:
fixup! revert unnecessary changes to binaryReader

rmmh commented:
The core idea of this PR is changing `query.RepoBranches{Set: map[string][]string}` into `query.RepoBranches{Set: map[string]roaring.Bitmap}` and `query.RepoSet{Set: map[string]bool}` into `query.RepoSet{IDs: *roaring.Bitmap}`, right? This has a chance of returning results from the wrong branch or repo on a hash collision, will the frontend perform the correct filtering in that case?

Commit:
reduce eval_test.go diff

Commit:
reduce diff of matchtree_test.go

Commit:
reduce diff in query.go and marshall.go

Commit:
do not RunOptimized in marshal function

Commit:
restore RepoBranches String method

Commit:
remove RepoBranches.Equal

Commit:
improved string methods

Commit:
1M repos in benchmarks

Commit:
5.5M benchmarks

tsenart commented:
`RepoBranches` benchmarks with 5.5M repos

```
name                       time/op
RepoBranches_Encode-12      1.34s ±13%
RepoBranches_Decode-12      1.21s ± 2%
RepoBranchesIDs_Encode-12  21.7ms ±40%
RepoBranchesIDs_Decode-12  16.6ms ± 2%

name                       bytes
RepoBranches_Encode-12      189MB ± 0%
RepoBranches_Decode-12
RepoBranchesIDs_Encode-12  12.8MB ± 0%
RepoBranchesIDs_Decode-12

name                       alloc/op
RepoBranches_Encode-12      189MB ± 0%
RepoBranches_Decode-12      766MB ± 0%
RepoBranchesIDs_Encode-12  14.2MB ± 0%
RepoBranchesIDs_Decode-12  33.1MB ± 0%

name                       allocs/op
RepoBranches_Encode-12       1.00 ± 0%
RepoBranches_Decode-12       134k ± 0%
RepoBranchesIDs_Encode-12    5.00 ± 0%
RepoBranchesIDs_Decode-12    172k ± 0%
```

tsenart commented:
> The core idea of this PR is changing query.RepoBranches{Set: map[string][]string} into query.RepoBranches{Set: map[string]roaring.Bitmap} and query.RepoSet{Set: map[string]bool} into query.RepoSet{IDs: *roaring.Bitmap}, right?

Yep!

> This has a chance of returning results from the wrong branch or repo on a hash collision, will the frontend perform the correct filtering in that case?

Not sure I understand. Can you elaborate on what you mean by a hash collision?

Commit:
Merge remote-tracking branch 'origin/master' into space-optimized-repo-sets

tsenart commented:
Alright, last commit, now with RepoSet benchmarks too:

```
name                       time/op
RepoBranches_Encode-12      1.37s ± 7%
RepoBranches_Decode-12      1.38s ±32%
RepoBranchesIDs_Encode-12  26.1ms ± 4%
RepoBranchesIDs_Decode-12  21.2ms ±23%
RepoSet_Encode-12           1.36s ±11%
RepoSet_Decode-12           1.55s ± 1%
RepoSetIDs_Encode-12       17.9ms ± 3%
RepoSetIDs_Decode-12       14.9ms ± 3%

name                       bytes
RepoBranches_Encode-12      189MB ± 0%
RepoBranches_Decode-12
RepoBranchesIDs_Encode-12  12.8MB ± 0%
RepoBranchesIDs_Decode-12
RepoSet_Encode-12           187MB ± 0%
RepoSet_Decode-12
RepoSetIDs_Encode-12       11.5MB ± 0%
RepoSetIDs_Decode-12

name                       alloc/op
RepoBranches_Encode-12      189MB ± 0%
RepoBranches_Decode-12      766MB ± 0%
RepoBranchesIDs_Encode-12  14.2MB ± 0%
RepoBranchesIDs_Decode-12  33.1MB ± 0%
RepoSet_Encode-12           226MB ± 0%
RepoSet_Decode-12           536MB ± 0%
RepoSetIDs_Encode-12       33.6MB ± 0%
RepoSetIDs_Decode-12       26.3MB ± 0%

name                       allocs/op
RepoBranches_Encode-12       1.00 ± 0%
RepoBranches_Decode-12       134k ± 0%
RepoBranchesIDs_Encode-12    5.00 ± 0%
RepoBranchesIDs_Decode-12    172k ± 0%
RepoSet_Encode-12           11.0M ± 0%
RepoSet_Decode-12           5.52M ± 0%
RepoSetIDs_Encode-12         8.00 ± 0%
RepoSetIDs_Decode-12         131k ± 0%
```

Commit:
RepoSet benchmarks

tsenart commented:
The PR should be a lot more "additive", so it's easier to review and later easier to delete deprecated code (i.e. `(RepoBranches
|RepoSet).Set)`)

keegancsmith commented on matchtree.go:
```diff
@@ -908,17 +908,26 @@ func (d *indexData) newMatchTree(q query.Q) (matchTree, error) {
+
+			switch {
+			case len(s.IDs) > 0:
+				for branch, ids := range s.IDs {
```
given we will delete `s.Set`, to make future PRs easier to read when we do that should we just always loop over s.IDs?

tsenart commented:
@keegancsmith: Yep, sequential IDs = massive gains, but more allocs. At some point we might want to contribute some optimizations on that front to roaring.

```
name                       old time/op    new time/op    delta
RepoBranches_Encode-12        1.37s ± 7%     1.37s ±13%       ~     (p=0.841 n=5+5)
RepoBranches_Decode-12        1.38s ±32%     1.27s ±17%       ~     (p=0.310 n=5+5)
RepoBranchesIDs_Encode-12    26.1ms ± 4%     0.2ms ± 9%    -99.35%  (p=0.008 n=5+5)
RepoBranchesIDs_Decode-12    21.2ms ±23%     0.3ms ±11%    -98.58%  (p=0.008 n=5+5)
RepoSet_Encode-12             1.36s ±11%     1.27s ±20%       ~     (p=0.151 n=5+5)
RepoSet_Decode-12             1.55s ± 1%     1.47s ±19%       ~     (p=0.190 n=4+5)
RepoSetIDs_Encode-12         17.9ms ± 3%     0.0ms ± 5%    -99.92%  (p=0.008 n=5+5)
RepoSetIDs_Decode-12         14.9ms ± 3%     0.0ms ± 2%    -99.73%  (p=0.008 n=5+5)

name                       old bytes      new bytes      delta
RepoBranches_Encode-12        189MB ± 0%     189MB ± 0%       ~     (all equal)
RepoBranchesIDs_Encode-12    12.8MB ± 0%     0.4MB ± 0%    -96.54%  (p=0.008 n=5+5)
RepoSet_Encode-12             187MB ± 0%     187MB ± 0%       ~     (all equal)
RepoSetIDs_Encode-12         11.5MB ± 0%     0.0MB ± 0%    -99.85%  (p=0.008 n=5+5)

name                       old alloc/op   new alloc/op   delta
RepoBranches_Encode-12        189MB ± 0%     189MB ± 0%       ~     (all equal)
RepoBranches_Decode-12        766MB ± 0%     766MB ± 0%       ~     (p=0.841 n=5+5)
RepoBranchesIDs_Encode-12    14.2MB ± 0%     0.5MB ± 0%       ~     (p=0.079 n=4+5)
RepoBranchesIDs_Decode-12    33.1MB ± 0%     0.9MB ± 0%    -97.22%  (p=0.008 n=5+5)
RepoSet_Encode-12             226MB ± 0%     226MB ± 0%       ~     (p=0.444 n=5+5)
RepoSet_Decode-12             536MB ± 0%     536MB ± 0%       ~     (p=0.421 n=5+5)
RepoSetIDs_Encode-12         33.6MB ± 0%     0.0MB ± 0%       ~     (p=0.079 n=4+5)
RepoSetIDs_Decode-12         26.3MB ± 0%     0.0MB ± 0%    -99.82%  (p=0.008 n=5+5)

name                       old allocs/op  new allocs/op  delta
RepoBranches_Encode-12         1.00 ± 0%      1.00 ± 0%       ~     (all equal)
RepoBranches_Decode-12         134k ± 0%      134k ± 0%       ~     (p=0.841 n=5+5)
RepoBranchesIDs_Encode-12      5.00 ± 0%     89.00 ± 0%  +1680.00%  (p=0.008 n=5+5)
RepoBranchesIDs_Decode-12      172k ± 0%        0k ± 0%    -99.74%  (p=0.008 n=5+5)
RepoSet_Encode-12             11.0M ± 0%     11.0M ± 0%       ~     (all equal)
RepoSet_Decode-12             5.52M ± 0%     5.52M ± 0%       ~     (p=0.421 n=5+5)
RepoSetIDs_Encode-12           8.00 ± 0%     87.00 ± 0%   +987.50%  (p=0.008 n=5+5)
RepoSetIDs_Decode-12           131k ± 0%        0k ± 0%    -99.66%  (p=0.008 n=5+5)
```

Commit:
benchmark with sequential IDs

rmmh commented:
As we discussed, the FNV hashing is only used for testing, otherwise we use repo IDs from our SQL backend.

The IDs generated for benchmarking shouldn't be *entirely* sequential, right? There are some gaps in the repo ids table (latest repoid is 56M but there are 2M active), so maybe make the generator have a random spacing in [0, 50]? The additional allocs are trivial and unconcerning-- an extra 80B of allocs here doesn't matter at all.

rmmh commented on matchtree.go:
```diff
@@ -908,17 +908,26 @@ func (d *indexData) newMatchTree(q query.Q) (matchTree, error) {
-						continue
+			var mask uint64
+
+			switch {
```
can this be simplified using s.Contains?

rmmh commented on query/query.go:
```diff
@@ -213,28 +234,61 @@ func (q *RepoBranches) Branches(name string) Q {
+}
+
+func (q *RepoSet) Contains(name string, id uint32) bool {
+	switch {
```
use if/else if

rmmh commented on query/marshal.go:
```diff
@@ -150,6 +144,69 @@ func repoBranchesDecode(b []byte) (map[string][]string, error) {
+		return uint64(binary.PutUvarint(enc[:], uint64(len(s))) + len(s))
+	}
+
+	// Calculate size
```
How much does this precise size calculation matter? I can see it slightly reducing garbage, but with the other changes it should be pretty minimal, and it takes longer to do the encoding twice like this.

keegancsmith replied:
It's likely not needed for IDs. For the older string based encoder it was very important due to the number of small objects.

rmmh reviewed (approved):
This looks good overall.

My primary nit throughout is that ```
switch {
case foo.Bar() > 0:
    ....
case baz.Quux() > 0:
    ....
}```
is weird compared to if/else if.

stefanhengl commented:
I was actually hoping we move away from RepoSets and RepoBranches, i.e. sending lists of repos around, and instead let Zoekt handle the raw query (for lack of a better word). Is there any reason why we absolutely have to send lists of repositories to Zoekt? @rvantonder is working on translating queries to Zoekt queries early, maybe we can build on that?

keegancsmith commented:
> I was actually hoping we move away from RepoSets and RepoBranches, i.e. sending lists of repos around, and instead let Zoekt handle the raw query (for lack of a better word). Is there any reason why we absolutely have to send lists of repositories to Zoekt? @rvantonder is working on translating queries to Zoekt queries early, maybe we can build on that?

We will likely always have lists of queries to send: eg repo groups or search contexts.

However, @tsenart and I chatted a bit about this. Why don't we just remove the known source of large lists. And we will! But we want to improve the encoding size anyways for any cases we accidently regress / places we have missed.

FYI the reposet work in this PR is actually unecessary. We don't use reposets anywhere on the client side, so it will never be marshalled.

keegancsmith reviewed (approved):
Some high level discussion we had over a call but not written down:

- I think it is safer if you introduced a new type. It feels dangerous that at every call site of a RepoBranches we now need to check which field is set. I'd be happy if you made that change before rolling this out, but I defer to you.
- The RepoSet change is unnecessary. We don't use them on the client side, so we never marshall the list. The only use of it currently is for type repo search, which is currently unused and likely could be better implemented when we do start needing it. I would prefer if you removed the RepoSet changes since we more than likely will delete RepoSet support.

tsenart commented on query/query.go:
```diff
@@ -197,9 +205,22 @@ func (q *RepoBranches) String() string {
+func (q *RepoBranches) Branches(name string, id uint32) Q {
+	var branches []string
+
+	switch {
```
Let's then agree to use if else for two branches, and switch for 2+ branches.

Commit:
address review feedback

Commit:
fixup! correct reason in docMatchTree

Commit:
Update query/marshal.go

Commit:
fixup! remove dead function

Commit:
fixup! refactor to use slice of tuples

Commit:
fixup! optimized ReposBranches encoding

name                     old time/op    new time/op    delta
RepoBranches_Encode-12      1.32s ± 7%     1.35s ± 5%     ~     (p=0.548 n=5+5)
RepoBranches_Decode-12      1.28s ±19%     1.30s ±25%     ~     (p=0.841 n=5+5)
BranchesRepos_Encode-12     511µs ±16%     160µs ± 8%  -68.61%  (p=0.008 n=5+5)
BranchesRepos_Decode-12     312µs ± 2%     289µs ± 2%   -7.31%  (p=0.008 n=5+5)

name                     old bytes      new bytes      delta
RepoBranches_Encode-12      189MB ± 0%     189MB ± 0%     ~     (all equal)
BranchesRepos_Encode-12     443kB ± 0%     443kB ± 0%   -0.00%  (p=0.008 n=5+5)

name                     old alloc/op   new alloc/op   delta
RepoBranches_Encode-12      189MB ± 0%     189MB ± 0%     ~     (all equal)
RepoBranches_Decode-12      766MB ± 0%     766MB ± 0%     ~     (p=0.651 n=5+5)
BranchesRepos_Encode-12    1.67MB ± 0%    0.45MB ± 0%  -72.93%  (p=0.008 n=5+5)
BranchesRepos_Decode-12     933kB ± 0%     919kB ± 0%     ~     (p=0.079 n=4+5)

name                     old allocs/op  new allocs/op  delta
RepoBranches_Encode-12       1.00 ± 0%      1.00 ± 0%     ~     (all equal)
RepoBranches_Decode-12       134k ± 0%      134k ± 0%     ~     (p=0.651 n=5+5)
BranchesRepos_Encode-12       109 ± 0%        90 ± 0%  -17.43%  (p=0.008 n=5+5)
BranchesRepos_Decode-12       820 ± 0%       447 ± 0%  -45.49%  (p=0.008 n=5+5)

Commit:
fixup! register BranchesRepos

Commit:
Update eval_test.go

Commit:
Update eval_test.go

Commit:
Update matchtree_test.go

Commit:
fixup! remove unnecessary transformers

Commit:
Revert "fixup! remove unnecessary transformers"

This reverts commit b2c60fb4f02276a44d735f99ad17ab7864145146.

tsenart commented:
I have introduced a new slice type `query.BranchesRepos` where each element is a tuple `(Branch, Repos)` and undid the changes to `query.RepoSet`.

keegancsmith committed:
make BranchesRepos a struct

Try and avoid footgun between type casting a query.Q between the value
vs pointer of BranchesRepos. Also consistent with the rest of query.Q
implementations.

keegancsmith commented:
Did some faking so we can run benchstat to compare RepoBranches vs the new BranchesRepos:
```
name                     old time/op    new time/op    delta
BranchesRepos_Encode-32     1.03s ± 7%     0.00s ± 3%    -99.99%  (p=0.008 n=5+5)
BranchesRepos_Decode-32     915ms ± 4%       0ms ± 1%    -99.98%  (p=0.008 n=5+5)

name                     old bytes      new bytes      delta
BranchesRepos_Encode-32     189MB ± 0%       0MB ± 0%    -99.77%  (p=0.008 n=5+5)

name                     old alloc/op   new alloc/op   delta
BranchesRepos_Encode-32     189MB ± 0%       0MB ± 0%    -99.76%  (p=0.000 n=4+5)
BranchesRepos_Decode-32     766MB ± 0%       1MB ± 0%    -99.88%  (p=0.008 n=5+5)

name                     old allocs/op  new allocs/op  delta
BranchesRepos_Encode-32      1.60 ±88%     89.00 ± 0%  +5462.50%  (p=0.008 n=5+5)
BranchesRepos_Decode-32      134k ± 0%        0k ± 0%    -99.67%  (p=0.008 n=5+5)
```

You can now pick up your jaw off the floor.

keegancsmith committed:
eval: avoid copying heavy RepoMetaData struct in simplification

keegancsmith committed:
matchtree: avoid copying Repositories in another repoMetaData loop

keegancsmith committed:
eval: remove special casing single repos to Branches

keegancsmith committed:
rpc: register branchesrepos as a pointer

keegancsmith committed:
inline single branch creation in selectRepoSet optimization

keegancsmith reviewed (approved):
shipit

keegancsmith committed:
Merge remote-tracking branch 'origin/master' into space-optimized-repo-sets
